### PR TITLE
Feat/main card

### PR DESCRIPTION
**Pull request by kacper-echaust**

### CODE REVIEW

**Review by Sumick — CHANGES_REQUESTED** [review]
I have 2 main issues with that code:

- [ ] You are encountering prop drilling problem, and it's getting messy to handle all the props that you pass below to your children components. Use react-context to prevent it and fix this problem by using it: https://react.dev/reference/react/createContext
- [ ] Second thing is that your code is very hard to read - please implement Prettier in your project and format your files: https://andrebnassis.medium.com/setting-prettier-on-a-react-typescript-project-2021-f9f0d5a1d6b0

**Review by NerdbordMentor — CHANGES_REQUESTED** [review]
## Code Review Report

### Task Goal
The task goal is to implement adding new flashcards with the following requirements:
1. Show a new card component when clicking on the `+` button.
2. Allow canceling the adding process.
3. Allow proceeding to the second side with a `Next` button.
4. Allow going back to the first side with a `Back` button.
5. Allow saving the new card.
6. Render the added card in the list when saved.

### Code Changes Review

#### 1. **src/App.css**
- Added font import and applied `Lexend` font family globally.
- **No issues found.**

#### 2. **src/App.jsx** (removed)
- The file was removed and replaced with `src/App.tsx`.

#### 3. **src/App.tsx** (added)
- Converted to TypeScript.
- Wrapped components with `CardListProvider` to provide context for card list management.
- **No issues found.**

#### 4. **src/components/AppHeader.module.css**
- Removed redundant font import and adjusted styles.
- **No issues found.**

#### 5. **src/components/AppHeader.tsx**
- Added context usage to manage the card list.
- Implemented `handleAddCard` function to add a new card when the `+` button is clicked.
- **No issues found.**

#### 6. **src/components/AppLayout.jsx** (removed)
- The file was removed and replaced with `src/components/AppLayout.tsx`.

#### 7. **src/components/AppLayout.tsx** (added)
- Converted to TypeScript.
- **No issues found.**

#### 8. **src/components/AppMain/AppMain.tsx** (added)
- Implemented to render `FaceCard` components based on the card list from context.
- **No issues found.**

#### 9. **src/components/AppMain/Card/FaceCard/FaceCard.module.css** (added)
- Added styles for the form, input, and buttons.
- **No issues found.**

#### 10. **src/components/AppMain/Card/FaceCard/FaceCard.tsx** (added)
- Implemented `FaceCard` component with input handling and buttons for `Cancel` and `Next`.
- **Issues found:**
  - The `Cancel` button does not have functionality to cancel the adding process.
  - The `Next` button does not proceed to the second side of the card.

#### 11. **src/components/Context/CardListProvider.tsx** (added)
- Implemented context provider for managing the card list state.
- **No issues found.**

#### 12. **src/components/types.ts** (added)
- Defined `CardList` type.
- **No issues found.**

#### 13. **src/global.d.ts** (renamed)
- No changes were made to this file.

#### 14. **src/main.jsx**
- Updated to use TypeScript version of `App`.
- **No issues found.**

### Summary
The code changes partially cover the task requirements:
1. **Show a new card component when clicking on the `+` button**: Implemented correctly.
2. **Allow canceling the adding process**: Not implemented.
3. **Allow proceeding to the second side with a `Next` button**: Not implemented.
4. **Allow going back to the first side with a `Back` button**: Not implemented.
5. **Allow saving the new card**: Not implemented.
6. **Render the added card in the list when saved**: Partially implemented (cards are added but not fully functional).

### Suggestions for Improvement
1. Implement functionality for the `Cancel` button in `FaceCard.tsx`:
    ```tsx
    const handleCancel = () => {
        setCardList(prevCardList => prevCardList.filter((_, i) => i !== index))
    }
    ```

2. Implement functionality for the `Next` button and add a second side:
    ```tsx
    const handleNext = () => {
        // Logic to proceed to the second side
    }
    ```

3. Implement functionality for saving the card:
    ```tsx
    const handleSave = () => {
        // Logic to save the card
    }
    ```

4. Add a state to manage which side of the card is being edited and implement `Back` button functionality.

By addressing these points, the implementation will fully meet the task requirements.

File: src/components/AppHeader.tsx [hunk 1]
```diff
@@ -1,16 +1,19 @@
...
 import styles from './AppHeader.module.css'
-type AppHeaderProps = {
-	cardsAmount: number
-}
-const AppHeader = ({ cardsAmount }: AppHeaderProps) => {
+import { CardContext } from './Context/CardListProvider'
+
+const AppHeader = () => {
+	const { cardList, setIsAdding } = useContext(CardContext)
+	const handleAddCard = () => {
```
Comment (Sumick) [inline src/components/AppHeader.tsx]: **nit:** This func name is quite misleading, as it does not add card but set card adding. I would rename it to `handleSetCardAdding`

File: src/components/AppMain/NewCard/NewCard.tsx [hunk 2]
```diff
@@ -0,0 +1,61 @@
+import { CardContext } from '../../Context/CardListProvider'
+import styles from './NewCard.module.css'
+import React, { ChangeEvent, useContext } from 'react'
+
+const NewCard = () => {
+	const { setEditingCard, editingCard, setIsAdding, isSecondSide, setIsSecondSide, setCardList } =
```
Comment (Sumick) [inline src/components/AppMain/NewCard/NewCard.tsx]: **nit:** setting card side should be a part of local state - not provider state.

File: src/components/Context/CardListProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+const CardContext = createContext<CardContextType>(defaultContext)
+
+const CardListProvider = ({ children }) => {
+	const [editingCard, setEditingCard] = useState<Card>({
+		title: '',
+		value: '',
+	})
+	const [cardList, setCardList] = useState<Card[]>([])
+	const [isAdding, setIsAdding] = useState(false)
+	const [isSecondSide, setIsSecondSide] = useState(false)
```
Comment (Sumick) [inline src/components/Context/CardListProvider.tsx]: **nit:** this should not be a part of the provider state, as it's related to the single card

File: src/components/Context/CardListProvider.tsx [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+	},
+	setEditingCard: () => {},
+	isSecondSide: false,
+	setIsSecondSide: () => {},
+}
+
+const CardContext = createContext<CardContextType>(defaultContext)
+
+const CardListProvider = ({ children }) => {
+	const [editingCard, setEditingCard] = useState<Card>({
```
Comment (Sumick) [inline src/components/Context/CardListProvider.tsx]: **nit**: this should not be a part of the provider state, as it's related to the single card

File: src/components/AppMain/NewCard/NewCard.tsx (hunk 2 above)
Comment (Sumick) [inline src/components/AppMain/NewCard/NewCard.tsx]: **nit:** you should set `currentSide` as string or enum to have a better control over your UI. So:

```tsx
const NewCard = () => {
  const [card, setCardValues] = useState<Card>({
                 title: '',
		value: '',
    })
  const [currentCardSide, setCardSide] = useState(CardSide.Front)

 const { handleAddCard  } = useContext(CardContext)
}
```

**Review by Sumick — CHANGES_REQUESTED** [review]
It's getting better, but we need to split logic that should be a part of provider state, and part of local card component state. Here you have an example - the rest is in comments below:

```tsx
const NewCard = () => {
  const [card, setCardValues] = useState<Card>({
                 title: '',
		value: '',
    })
  const [currentCardSide, setCardSide] = useState(CardSide.Front)
 const { handleAddCard  } = useContext(CardContext)
}
```

In the context you should keep your "global" data, like all added cards, and logic for adding card. All other things like controlling component inside this component should be inside that component.

**Review by Sumick — APPROVED** [review]
Looks good! Im still having problem with that formatting - can you check please, this specific command? Run `npx prettier --write .` from the root of your repository, then do `git add .`

**Pull request merged**